### PR TITLE
ActiveRecord, Direct Queries and rake build tasks

### PR DESCRIPTION
**Pull request by rodrigovdb**
# ActiveRecord

Adds Sinatra's ActiveRecord gem and maps OrderByDay and HitByDay table.
Also creates methods to search informations to dashboard, orders by source and tracker.
All these methods are covered by tests and don't return hits data yet.
# Rake tasks

Creates rake tasks to:
- Create all tables
- Import hits data from wiseleph_hattan database
- IMport orders data from tusks
- Log all imports

These tasks use a collect of methods to abstract database connection and allow run SQL queries
directly from code.
# Controller routes and model methods

Creates routes to dashboard, orders by source and tracker reports.

### CODE REVIEW

File: app/models/orders_by_day.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    response[:tracker_url]  = item.tracker_url
+    response.merge! sumarize(items)
+
+    response.to_json
+  end
+
+  def self.assemble_sources_hash(items)
+    response  = {}
+    split     = {}
+    template  = { hits: 0, conversions: 0, avg_order_value: 0 }
```
Comment (fellipebrito) [inline app/models/orders_by_day.rb:38]: This "template" idea is neat! :clap:

File: lib/tasks/create_tables.rake [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    include Report::MysqlConnection
+
+    create_hits_by_day reports_client
+    create_orders_by_day reports_client
+    create_import_logs reports_client
+  end
+end
+
+def create_hits_by_day(client)
+    query =<<EOF
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:13]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/create_tables.rake [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY,
+  created_at date DEFAULT NULL,
+  tracker_name varchar(70) CHARACTER SET utf8 DEFAULT NULL,
+  destination_name varchar(255) CHARACTER SET utf8 DEFAULT NULL,
+  hits INTEGER NOT NULL DEFAULT -1
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
+EOF
+    client.query query
+
+    query = "CREATE INDEX reports_hits_by_day_created_at_idx ON reports.hits_by_day(created_at) USING BTREE;"
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:24]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/create_tables.rake [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    client.query query
+
+    query = "CREATE INDEX reports_hits_by_day_created_at_tracker_name_idx ON reports.hits_by_day(created_at, tracker_name) USING BTREE"
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:27]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/create_tables.rake [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    client.query query
+end
+
+def create_orders_by_day(client)
+    query =<<EOF
+CREATE TABLE orders_by_day (
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:33]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/create_tables.rake [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  platform_url varchar(1000) DEFAULT NULL,
+  source_name varchar(1000) DEFAULT NULL,
+  source_display_name varchar(255) NOT NULL,
+  number_of_orders bigint(21) NOT NULL,
+  value_of_orders double NOT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
+EOF
+    client.query query
+
+    query = 'CREATE INDEX reports_order_by_day_order_type_idx ON reports.orders_by_day(order_type) USING BTREE;';
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:53]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/create_tables.rake [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    client.query query
+
+    query = 'CREATE INDEX reports_order_by_day_destination_id_idx ON reports.orders_by_day(destination_id) USING BTREE;';
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:56]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/create_tables.rake [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    client.query query
+
+    query = 'CREATE INDEX reports_order_by_day_tracker_id_idx ON reports.orders_by_day(tracker_id) USING BTREE;';
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:59]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/create_tables.rake [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    client.query query
+end
+
+def create_import_logs(client)
+    query =<<EOF
+CREATE TABLE reports.daily_import_logs(
```
Comment (fellipebrito) [inline lib/tasks/create_tables.rake:65]: I'd rather keep this query into a separated file under /sql folder. It is easy to maintain and change. You do not have to do it now. But keep it in mind for a future refactor.

File: lib/tasks/mysql_connection_helper.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
+require 'mysql2'
+require 'yaml'
+
+# Require this on your rake tasks and everything will be fine!
+# Use require Report::MysqlConnection and use reports_client as well.
+module Report
```
Comment (fellipebrito) [inline lib/tasks/mysql_connection_helper.rb:6]: This module should be an open source code @ devlandia... it is really good. It needs one refactor here and there so we can use dynamic connections, but it is great as it is.

**Comment by fellipebrito** [comment]
There are some other possible refactors here in code climate https://codeclimate.com/repos/5548e027e30ba04d06019cbc/compare/active_record#ratings

I'd consider taking a look at it and then examine your schedule and manage what to do in a way that does not compromise your delivery date.

File: lib/tasks/create_tables.rake (hunk 2 above)
Comment (rodrigovdb) [inline lib/tasks/create_tables.rake:13]: Sounds like a great idea!

**Pull request merged**